internal/promote/storage.go

```diff
@@ -0,0 +1,410 @@
+		}
+
+		c, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading file content failed (path: %s)", path)
+		}
+
+		m[path] = c
+	}
+	return m, nil
+}
+
+func writePackageContents(filesystem billy.Filesystem, contents fileContents) error {
+	for resourcePath, content := range contents {
+		dir := filepath.Dir(resourcePath)
+		err := filesystem.MkdirAll(dir, 0644)
+		if err != nil {
+			return errors.Wrapf(err, "creating directory failed (path: %s)", dir)
+		}
+
+		err = util.WriteFile(filesystem, resourcePath, content, 0755)
+		if err != nil {
+			return errors.Wrapf(err, "writing file failed (path: %s)", dir)
+		}
+	}
+	return nil
+}
+
+// RemovePackages method removes packages from "stage" branch. It creates a new branch with removed packages.
+func RemovePackages(r *git.Repository, sourceStage string, packages PackageRevisions, nonce int64) (string, error) {
+	fmt.Printf("Remove packages from %s...\n", sourceStage)
+
+	wt, err := r.Worktree()
+	if err != nil {
+		return "", errors.Wrap(err, "fetching worktree reference failed")
+	}
+
+	// Create branch for updated stage
+	err = wt.Checkout(&git.CheckoutOptions{
+		Branch: plumbing.NewBranchReferenceName(sourceStage),
+	})
+	if err != nil {
+		return "", errors.Wrapf(err, "changing branch failed (path: %s)", sourceStage)
+	}
+
+	newSourceStage := fmt.Sprintf("delete-from-%s-%d", sourceStage, nonce)
+	err = wt.Checkout(&git.CheckoutOptions{
+		Branch: plumbing.NewBranchReferenceName(newSourceStage),
+		Create: true,
+	})
+	if err != nil {
+		return "", errors.Wrapf(err, "changing branch failed (path: %s)", newSourceStage)
+	}
+
+	for _, p := range packages {
+		_, err := wt.Remove(p.path())
+		if err != nil {
+			return "", errors.Wrapf(err, "removing package from index failed (path: %s)", p.path())
+		}
+	}
+
+	_, err = wt.Commit(fmt.Sprintf("Delete packages from %s", sourceStage), new(git.CommitOptions))
+	if err != nil {
+		return "", errors.Wrapf(err, "committing files failed (stage: %s)", sourceStage)
+	}
+	return newSourceStage, nil
+}
+
+// PushChanges method pushes all branches to the remote repository.
+func PushChanges(r *git.Repository) error {
+	user, err := User(r)
+	if err != nil {
+		return errors.Wrap(err, "reading Git user failed")
+	}
+
+	authToken, err := github.AuthToken()
+	if err != nil {
+		return errors.Wrap(err, "reading auth token failed")
+	}
+
+	err = r.Push(&git.PushOptions{
+		RemoteName: user,
+		Auth: &http.BasicAuth{
+			Username: user,
+			Password: authToken,
+		},
+	})
+	if err != nil {
+		return errors.Wrap(err, "pushing branch failed")
+	}
+	return nil
+}
+
+// User method returns the Git user.
+func User(r *git.Repository) (string, error) {
+	c, err := r.ConfigScoped(config.SystemScope)
+	if err != nil {
+		return "", errors.Wrap(err, "reading config failed")
+	}
+	return c.User.Name, nil
+}
+
+func sortPackageRevisions(revisions PackageRevisions) PackageRevisions {
```

Comment:
Might be personal preference but when I see a function whose name contains a type in it and also takes an argument of the same type, it feels like the function should really be a method with the type as the receiver. So something like:

```go
func (p PackageRevisions) sort() PackageRevisions {
   ...
}
```

Reply:
I like this idea, it's better than simple producedure. Fixed.